src/persist/src/s3.rs

```diff
@@ -163,30 +163,216 @@ impl S3BlobCore {
+            let mut min_header_elapsed = MinElapsed::default();
+            for part_fut in part_futs {
+                let (this_header_elapsed, part_res) = part_fut
+                    .await
+                    .map_err(|err| Error::from(format!("s3 spawn err: {}", err)))?;
+                let part_res =
+                    part_res.map_err(|err| Error::from(format!("s3 get meta err: {}", err)))?;
+                let this_body_len = usize::try_from(part_res.content_length)
+                    .map_err(|err| format!("unexpected s3 content_length: {}", err))?;
+                total_body_len += this_body_len;
+
+                let min_header_elapsed = min_header_elapsed.observe(this_header_elapsed);
+                if this_header_elapsed >= min_header_elapsed * 8 {
+                    tracing::debug!(
+                        "s3 download part header took {:?} more than 8x the min {:?}",
+                        this_header_elapsed,
+                        min_header_elapsed
+                    );
+                } else {
+                    tracing::trace!("s3 download part header took {:?}", this_header_elapsed);
+                }
+
+                let start_part_body = Instant::now();
+                part_bodies.push(
+                    async_runtime.spawn_named(
+                        || "persist_s3blob_get_body",
+                        part_res
+                            .body
+                            .collect()
+                            .map(move |res| (start_part_body.elapsed(), this_body_len, res)),
+                    ),
+                );
+            }
+            tracing::trace!(
+                "s3 download part headers took {:?} ({} parts)",
+                start_headers.elapsed(),
+                num_parts
+            );
+        }
+
+        let start_bodies = Instant::now();
+        let mut min_body_elapsed = MinElapsed::default();
+        let mut val = vec![0u8; total_body_len];
+        let mut val_offset = 0;
+        let mut body_copy_elapsed = Duration::ZERO;
+        for part_body in part_bodies {
+            let (this_body_elapsed, this_body_len, part_body_res) = part_body
+                .await
+                .map_err(|err| Error::from(format!("s3 spawn err: {}", err)))?;
+            let mut part_body_res =
+                part_body_res.map_err(|err| Error::from(format!("s3 get body err: {}", err)))?;
+
+            let min_body_elapsed = min_body_elapsed.observe(this_body_elapsed);
+            if this_body_elapsed >= min_body_elapsed * 8 {
+                tracing::debug!(
+                    "s3 download part body took {:?} more than 8x the min {:?}",
+                    this_body_elapsed,
+                    min_body_elapsed
+                );
+            } else {
+                tracing::trace!("s3 download part body took {:?}", this_body_elapsed);
+            }
+
+            // TODO: It'd be lovely if we could push these copies into the
```

Comment:
It's probably worth doing this TODO sooner than later. a 6GB blob takes ~6s to download and ~3s to collect into a single Vec<u8>. plus all the copy_to_slice math feels brittle